src/World.cpp

```diff
@@ -572,50 +572,147 @@ void cWorld::Start(void)
+	} while ((--BiomeCheckCount) > 0);
+	
+	// Check 0, 0 first.
+	double SpawnY = 0.0;
+	if (CanSpawnAt(BiomeOffset.x, BiomeOffset.z, SpawnY))
+	{
+		m_SpawnX = BiomeOffset.x + 0.5;
+		m_SpawnY = SpawnY;
+		m_SpawnZ = BiomeOffset.z + 0.5;
+
+		LOGINFO("Generated spawnpoint position at {%i, %i, %i}", static_cast<int>(m_SpawnX), static_cast<int>(m_SpawnY), static_cast<int>(m_SpawnZ));
+		return;
+	}
+
+	// A search grid (searches clockwise around the origin)
+	static const Vector3i ChunkOffset[] =
+	{
+		Vector3i(0, 0, cChunkDef::Width),
+		Vector3i(cChunkDef::Width, 0, cChunkDef::Width),
+		Vector3i(cChunkDef::Width, 0, 0),
+		Vector3i(cChunkDef::Width, 0, -cChunkDef::Width),
+		Vector3i(0, 0, -cChunkDef::Width),
+		Vector3i(-cChunkDef::Width, 0, -cChunkDef::Width),
+		Vector3i(-cChunkDef::Width, 0, 0),
+		Vector3i(-cChunkDef::Width, 0, cChunkDef::Width),
+	};
+
+	static const int PerRadiSearchCount = ARRAYCOUNT(ChunkOffset);
+
+	for (int RadiusOffset = 1; RadiusOffset < a_MaxSpawnRadius; ++RadiusOffset)
+	{
+		for (int SearchGridIndex = 0; SearchGridIndex < PerRadiSearchCount; ++SearchGridIndex)
+		{
+			const Vector3i PotentialSpawn = BiomeOffset + (ChunkOffset[SearchGridIndex] * RadiusOffset);
+			
+			if (CanSpawnAt(PotentialSpawn.x, PotentialSpawn.z, SpawnY))
+			{
+				m_SpawnX = PotentialSpawn.x + 0.5;
+				m_SpawnY = SpawnY;
+				m_SpawnZ = PotentialSpawn.z + 0.5;
+
+				LOGINFO("Generated spawnpoint position at {%i, %i, %i}", static_cast<int>(m_SpawnX), static_cast<int>(m_SpawnY), static_cast<int>(m_SpawnZ));
+				return;
+			}
+		}
+	}
+
+	m_SpawnY = GetHeight(static_cast<int>(m_SpawnX), static_cast<int>(m_SpawnZ));
+	LOGWARNING("Did not find an acceptable spawnpoint. Generated a random spawnpoint position at {%i, %i, %i}", static_cast<int>(m_SpawnX), static_cast<int>(m_SpawnY), static_cast<int>(m_SpawnZ));
+}
+
+
+
+
+
+bool cWorld::CanSpawnAt(double a_X, double a_Z, double &a_Y)
+{
+	static const BLOCKTYPE ValidSpawnBlocks[] =
+	{
+		E_BLOCK_GRASS,
+		E_BLOCK_DIRT,
+		E_BLOCK_SAND,
+		E_BLOCK_SNOW,
+		E_BLOCK_SNOW_BLOCK,
+		E_BLOCK_NETHERRACK
+	};
+
+	static const int ValidSpawnBlocksCount = ARRAYCOUNT(ValidSpawnBlocks);
+
+	for (int PotentialY = 96; PotentialY > 32; --PotentialY)
+	{
+		BLOCKTYPE HeadBlock = GetBlock(static_cast<int>(a_X), PotentialY, static_cast<int>(a_Z));
+
+		// Is this block safe for spawning
+		if (HeadBlock != E_BLOCK_AIR)
+		{
+			continue;
+		}
+
+		BLOCKTYPE BodyBlock = GetBlock(static_cast<int>(a_X), PotentialY - 1, static_cast<int>(a_Z));
+
+		// Is this block safe for spawning
+		if (BodyBlock != E_BLOCK_AIR)
+		{
+			continue;
+		}
+
+		BLOCKTYPE FloorBlock = GetBlock(static_cast<int>(a_X), PotentialY - 2, static_cast<int>(a_Z));
+
+		// Is the floor block ok
+		bool ValidSpawnBlock = false;
+		for (int BlockIndex = 0; BlockIndex < ValidSpawnBlocksCount; ++BlockIndex)
 		{
-			if (CheckPlayerSpawnPoint(SpawnX, GetHeight(SpawnX, SpawnZ), SpawnZ))
+			if (ValidSpawnBlocks[BlockIndex] == FloorBlock)
```

Comment:
Replace this if with a ternary operator which allows the compiler to unroll the loop. This loop body is going to be executed 61440 times.